
some of these hacks  should be solved properly by upstream ...

--- glibc-2.9.orig/nscd/connections.c
+++ glibc-2.9/nscd/connections.c
@@ -873,6 +873,7 @@
 #endif
       }
 
+#if defined(SOCK_CLOEXEC) && defined(SOCK_NONBLOCK)
   /* Create the socket.  */
 #ifndef __ASSUME_SOCK_CLOEXEC
   sock = -1;
@@ -885,6 +886,7 @@
 	have_sock_cloexec = sock != -1 || errno != EINVAL ? 1 : -1;
 #endif
     }
+#endif    
 #ifndef __ASSUME_SOCK_CLOEXEC
   if (have_sock_cloexec < 0)
     sock = socket (AF_UNIX, SOCK_STREAM, 0);
--- glibc-2.9.orig/sysdeps/i386/tlsdesc.sym
+++ glibc-2.9/sysdeps/i386/tlsdesc.sym
@@ -8,7 +8,11 @@
 
 -- Abuse tls.h macros to derive offsets relative to the thread register.
 
+#if 0
 DTV_OFFSET			offsetof(struct pthread, header.dtv)
+#else
+DTV_OFFSET                      offsetof(struct _pthread_descr_struct, p_header.data.dtvp)
+#endif
 
 TLSDESC_ARG			offsetof(struct tlsdesc, arg)
 
--- glibc-2.9.orig/sysdeps/x86_64/tlsdesc.sym
+++ glibc-2.9/sysdeps/x86_64/tlsdesc.sym
@@ -5,10 +5,13 @@
 #include <dl-tlsdesc.h>
 
 --
+#if 0
+ DTV_OFFSET			offsetof(struct pthread, header.dtv) */
+#endif
 
 -- Abuse tls.h macros to derive offsets relative to the thread register.
 
-DTV_OFFSET			offsetof(struct pthread, header.dtv)
+DTV_OFFSET			offsetof(struct _pthread_descr_struct, p_header.data.dtvp)
 
 TLSDESC_ARG			offsetof(struct tlsdesc, arg)
 
--- glibc-2.9.orig/linuxthreads/sysdeps/pthread/malloc-machine.h
+++ glibc-2.9/linuxthreads/sysdeps/pthread/malloc-machine.h
@@ -57,10 +57,10 @@
 #include <bits/libc-tsd.h>
 
 typedef int tsd_key_t[1];	/* no key data structure, libc magic does it */
-__libc_tsd_define (static, MALLOC)	/* declaration/common definition */
+__libc_tsd_define (static, void *, MALLOC)	/* declaration/common definition */
 #define tsd_key_create(key, destr)	((void) (key))
-#define tsd_setspecific(key, data)	__libc_tsd_set (MALLOC, (data))
-#define tsd_getspecific(key, vptr)	((vptr) = __libc_tsd_get (MALLOC))
+#define tsd_setspecific(key, data)	__libc_tsd_set (void *, MALLOC, (data))
+#define tsd_getspecific(key, vptr)	((vptr) = __libc_tsd_get (void *, MALLOC))
 
 #include <sysdeps/generic/malloc-machine.h>
 
